### ISSUE TITLE
Potential bugs when mmdetection runs on PyTorch < 1.8

### ISSUE THREAD

**Issue by serend1p1ty** [open-mmlab/mmdetection#8110]
Thanks for your error report and we appreciate it a lot.

**Describe the bug**

There is a small [bug](https://github.com/pytorch/pytorch/issues/5059) of  PyTorch (version < 1.8).

**In short, when we use `DataLoadr` with `num_workers != 0`, these forked children process share the same `numpy` random seeds.**

A toy example, run with PyTorch 1.6.

```
import numpy as np
from torch.utils.data import Dataset, DataLoader

class RandomDataset(Dataset):
    def __getitem__(self, index):
        return np.random.randint(0, 1000, 3)

    def __len__(self):
        return 8

dataset = RandomDataset()
dataloader = DataLoader(dataset, batch_size=2, num_workers=2)
for batch in dataloader:
    print(batch)
```

Output:

```
tensor([[437, 650, 998],  # process 0
        [108,  34, 197]])
tensor([[437, 650, 998],  # process 1
        [108,  34, 197]])
tensor([[153, 629, 103],  # process 0
        [695, 102, 728]])
tensor([[153, 629, 103],  # process 1
        [695, 102, 728]])
```

Currently, mmdetection sets different `numpy` seeds for different workers according to the `worker_id`. However, the worker processes are killed at the end of each epoch, and all worker resources are lost. At the next epoch, mmdetection will set the same seeds as the previous epoch.

Example:

```
import numpy as np
from torch.utils.data import Dataset, DataLoader

class RandomDataset(Dataset):
    def __getitem__(self, index):
        return np.random.randint(0, 1000, 3)

    def __len__(self):
        return 8

def worker_init_fn(worker_id):
    np.random.seed(worker_id)

dataset = RandomDataset()
dataloader = DataLoader(dataset, batch_size=2, num_workers=2, worker_init_fn=worker_init_fn)

print("the first epoch is ok")
for batch in dataloader:
    print(batch)

print("the second epoch is the same as the first epoch")
for batch in dataloader:
    print(batch)
```

Output:

```
the first epoch is ok

tensor([[684, 559, 629],
        [192, 835, 763]])
tensor([[ 37, 235, 908],
        [ 72, 767, 905]])
tensor([[707, 359,   9],
        [723, 277, 754]])
tensor([[715, 645, 847],
        [960, 144, 129]])

the second epoch is the same as the first epoch

tensor([[684, 559, 629],
        [192, 835, 763]])
tensor([[ 37, 235, 908],
        [ 72, 767, 905]])
tensor([[707, 359,   9],
        [723, 277, 754]])
tensor([[715, 645, 847],
        [960, 144, 129]])
```

Maybe we can use the solution of [detectron2](https://github.com/facebookresearch/detectron2/blob/523c4023bc79719d2fc3fe0181311f88ad35a271/detectron2/data/build.py#L554) or [pytorch-image-models](https://github.com/rwightman/pytorch-image-models/blob/e4360e6125bb0bb4279785810c8eb33b40af3ebd/timm/data/loader.py#L149).

Relevent issue: https://github.com/pytorch/pytorch/issues/5059

**Environment**

PyTorch 1.6.0

**Comment by jbwang1997** [comment]
Thank you very much for your error report. In our case, we use a custom sampler when building dataloader. In the sampler, we set different seeds to each epoch and rank.

The code can refer to 
https://github.com/open-mmlab/mmdetection/blob/master/mmdet/datasets/samplers/distributed_sampler.py#L11-L54

**Comment by serend1p1ty** [comment]
Actually, the problem is not related to distributed training (e.g., distributed_sampler)

As long as `num_workers != 0`, we will see duplicate data during training.

Please see the following example using `mmdet.datasets.builder.build_dataloader`.

```
import numpy as np
import mmdet.datasets.builder as builder

class RandomDataset(Dataset):
    def __getitem__(self, index):
        return np.random.randint(0, 1000, 3)

    def __len__(self):
        return 8

dataset = RandomDataset()
dataloader = builder.build_dataloader(dataset, samples_per_gpu=2, workers_per_gpu=2, shuffle=False)

for batch in dataloader:
    print(batch)
```

Output:

```
tensor([[301, 462, 262],
        [195, 738, 766]])
tensor([[301, 462, 262],
        [195, 738, 766]])
tensor([[347, 805, 531],
        [417, 946, 716]])
tensor([[347, 805, 531],
        [417, 946, 716]])
```

**Comment by jbwang1997** [comment]
I reproduce the same problem according to this issue. And I have made a pr to fix it. Thank you again for this bug report.

**Comment by hhaAndroid** [comment]
@serend1p1ty Thanks a lot for your feedback. In fact, this problem exists in all PyTorch versions, and [the official solution](https://pytorch.org/docs/stable/notes/randomness.html#dataloader) has been given.

 - no problem with spawn mode
 - the `persistent_workers` parameter was introduced in pytorch1.7, and this problem can be avoided as long as it is set to True.  But if you set it to false it will still appear

**Comment by serend1p1ty** [comment]
> @serend1p1ty Thanks a lot for your feedback. In fact, this problem exists in all PyTorch versions, and [the official solution](https://pytorch.org/docs/stable/notes/randomness.html#dataloader) has been given.
> 
> * no problem with spawn mode
> * the `persistent_workers` parameter was introduced in pytorch1.7, and this problem can be avoided as long as it is set to True.  But if you set it to false it will still appear

- Actually, not all PyTorch versions have this problem. After 1.9, the official [has fixed it](https://github.com/pytorch/pytorch/blob/cf3ce329b548e1e8da65542af53b16f27d5dc72a/torch/utils/data/_utils/worker.py#L220).
- Yes, no problem with `spawn` mode. However, according to the [offcial docs](https://pytorch.org/docs/stable/data.html#platform-specific-behaviors), `torch.multiprocessing` use the `fork` as the default method on Unix.
- Even use `persistent_workers`, we still need to set a random seed for numpy in the first epoch. And this method is not compatible with earlier versions of PyTorch (emmm, I think not graceful enough). The official solution using `torch.initial_seed()` is a better choice.
